### **Parte 1:** Importanción de librerías

#### Suprimiendo los Warnings

In [3]:
import warnings
warnings.filterwarnings("ignore")

#### Importando las librerías para trabajar CNN

In [58]:
from keras.models import Sequential
from keras.layers import Convolution2D        
from keras.layers import MaxPooling2D       
from keras.layers import Flatten            
from keras.layers import Dense 

from keras.preprocessing.image import ImageDataGenerator

from tensorflow.keras.preprocessing import image
import numpy as np

### **Parte 2:** Construcción de la CNN

In [5]:
#Se crea y se inicializa la CNN
classifier = Sequential()

#Paso 1 - Convolución
classifier.add(Convolution2D(
    filters=32,
    kernel_size=(3,3),
    input_shape = (64,64,3),
    activation="relu"
))

#Paso 2 - Max Pooling
classifier.add(MaxPooling2D(
    pool_size=(2,2)
))

#Se agrega otra capa de convolución y de max pooling para mejorar la red
classifier.add(Convolution2D(
    filters=32,
    kernel_size=(3,3),
    activation="relu"
))

classifier.add(MaxPooling2D(
    pool_size=(2,2)
))

#Paso 3 - Flattening
classifier.add(Flatten())

#Paso 4 - Full Connection
classifier.add(Dense(
    units=128,
    activation="relu"
))

classifier.add(Dense(
    units=1,
    activation="sigmoid"
))

#Compilación de la CNN
classifier.compile(
    optimizer = "adam",
    loss = "binary_crossentropy",
    metrics = ["accuracy"]
)

### **Parte 3:** Ajustar la imágenes para entrenar la CNN

In [6]:
train_datagen = ImageDataGenerator(
    rescale=1./255,
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True
)

test_datagen = ImageDataGenerator(
    rescale=1./255
)

training_dataset = train_datagen.flow_from_directory(
    directory='dataset/training_set/',
    target_size=(64, 64),
    batch_size=40,
    class_mode='binary'
)

testing_dataset = test_datagen.flow_from_directory(
    directory='dataset/test_set/',
    target_size=(64, 64),
    batch_size=40,
    class_mode='binary'
)

classifier.fit_generator(
    training_dataset,
    steps_per_epoch=200,
    epochs=25,
    validation_data=testing_dataset,
    validation_steps=50
)

Found 8000 images belonging to 2 classes.
Found 2000 images belonging to 2 classes.
Epoch 1/25
200/200 [==============================] - 111s 549ms/step - loss: 0.6847 - accuracy: 0.5541 - val_loss: 0.6297 - val_accuracy: 0.6640
Epoch 2/25
200/200 [==============================] - 88s 440ms/step - loss: 0.6208 - accuracy: 0.6555 - val_loss: 0.6324 - val_accuracy: 0.6430
Epoch 3/25
200/200 [==============================] - 102s 508ms/step - loss: 0.5816 - accuracy: 0.6986 - val_loss: 0.6299 - val_accuracy: 0.6625
Epoch 4/25
200/200 [==============================] - 119s 589ms/step - loss: 0.5384 - accuracy: 0.7266 - val_loss: 0.5724 - val_accuracy: 0.7065
Epoch 5/25
200/200 [==============================] - 87s 435ms/step - loss: 0.5241 - accuracy: 0.7399 - val_loss: 0.4962 - val_accuracy: 0.7675
Epoch 6/25
200/200 [==============================] - 80s 399ms/step - loss: 0.4901 - accuracy: 0.7638 - val_loss: 0.4882 - val_accuracy: 0.7790
Epoch 7/25
200/200 [=======================

### **Parte 4:** Realizar una simple predicción

In [59]:
#Carga la imagen de prueba
imagen_prueba = image.load_img(
    'dataset/single_prediction/cat_or_dog_1.jpg',
    target_size = (64, 64)
)

#Convierte la imagen a un array numpy
imagen_array = image.img_to_array(imagen_prueba)
imagen_array = np.expand_dims(imagen_array, axis=0)

# Normalizar la imagen
imagen_array /= 255.0

#Realizar la predicción
prediccion = classifier.predict(imagen_array)

1/1 [==============================] - 0s 121ms/step


A continuación, se procede a verificar cómo están etiquetadas las clases.

In [64]:
training_dataset.class_indices

{'cats': 0, 'dogs': 1}

* Si la probabilidad predicha es cercana a 1, el modelo está prediciendo con alta confianza la clase etiquetada como 1 ('dogs').
* Si la probabilidad predicha es cercana a 0, el modelo está prediciendo con alta confianza la clase etiquetada como 0 ('cats').

Verificando la predicción

In [65]:
prediccion[0][0]

0.99999756

**Conclusión:** Dado que el resultado de la predicción es 0.99999756 y la clase 'dogs' ('perros') 
está etiquetada como 1, puedes interpretar esa predicción como altamente indicativa 
de que la imagen pertenece a la clase 'dogs' ('perros'). En este caso, una probabilidad 
cercana a 1 sugiere una alta confianza del modelo en que la imagen es un perro según la 
configuración de etiquetas y clases.